### ISSUE TITLE
Heredoc delimiter/marker not parsed correctly

### ISSUE THREAD

**Issue by Dabombber** [koalaman/shellcheck#1650]
Heredocs made by perverse people where the delimiter contains the `'` or `"` characters are not parsed correctly. I'm not sure if it's valid code, but it works in Busybox's ash and Bash. The [posix docs](http://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_07_04) say
> If any part of word is quoted, the delimiter shall be formed by performing quote removal on word, and the here-document lines shall not be expanded.

#### For bugs
- Rule Ids: SC1073, SC1072, SC1042, SC1041
- version: 0.6.0
- [x] The rule's wiki page does not already cover this (e.g. https://shellcheck.net/wiki/SC1041)
- [x] I tried on shellcheck.net and verified that this is still a problem on the latest commit

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/bin/sh

cat <<'"EOF'
hello
"EOF
```

likewise for 
```sh
#!/bin/sh

cat <<'E'\''OF'
hello
E'OF
```

#### Here's what shellcheck currently says:

> Line 3:
> cat <<'"EOF'
> ^-- SC1009: The mentioned syntax error was in this simple command.
>     ^-- SC1073: Couldn't parse this here document. Fix to allow more checks.
>       ^-- SC1041: Found 'EOF' further down, but not on a separate line.
>       ^-- SC1042: Close matches include '"EOF' (!= 'EOF').
>  
> Line 6:
> 
> ^-- SC1072: Here document was not correctly terminated. Fix any mentioned problems and try again.

#### Here's what I wanted or expected to see:

> No issues detected!
